transport/tlsfrag/buffer.go

```diff
@@ -0,0 +1,134 @@
+	errInvalidTLSClientHello = errors.New("not a valid TLS Client Hello packet")
+)
+
+// clientHelloBuffer is a byte buffer used to receive and buffer a TLS Client Hello packet.
+type clientHelloBuffer struct {
+	data    []byte        // The buffer that hosts both header and content, cap: 5 -> 5+len(content)+padding
+	padding int           // The unused additional padding allocated at the end of data, 0 -> 5
+	valid   bool          // Indicates whether the content in data is a valid TLS Client Hello record
+	bufrd   *bytes.Reader // A reader used to read from the slice passed to Write
+}
+
+var _ io.Writer = (*clientHelloBuffer)(nil)
+var _ io.ReaderFrom = (*clientHelloBuffer)(nil)
+
+// newClientHelloBuffer creates and initializes a new buffer to receive a TLS Client Hello packet.
+func newClientHelloBuffer() *clientHelloBuffer {
+	// Allocate the 5 bytes header first, and then reallocate it to contain the entire packet later
+	return &clientHelloBuffer{
+		data:    make([]byte, 0, recordHeaderLen),
+		padding: 0,
+		valid:   true,
+		bufrd:   bytes.NewReader(nil), // It will be Reset in Write
+	}
+}
+
+// Bytes returns the full Client Hello packet including both the 5 bytes header and the content.
+func (b *clientHelloBuffer) Bytes() []byte {
+	return b.data
+}
+
+// Write appends p to the buffer and returns the number of bytes actually used.
+// If this data completes a valid TLS Client Hello, it returns errTLSClientHelloFullyReceived.
+// If an invalid TLS Client Hello message is detected, it returns the error errInvalidTLSClientHello.
+// If all bytes in p have been used and the buffer still requires more data to build a complete TLS Client Hello
+// message, it returns (len(p), nil).
+func (b *clientHelloBuffer) Write(p []byte) (int, error) {
+	b.bufrd.Reset(p)
+	n, err := b.ReadFrom(b.bufrd)
+	if err == nil && int(n) != len(p) {
+		err = io.ErrShortWrite
+	}
+	return int(n), err
+}
+
+// ReadFrom reads all the data from r and appends it to this buffer until a complete Client Hello packet has been
+// received, or r returns EOF or error. It returns the number of bytes read. Any error except EOF encountered during
+// the read is also returned.
+//
+// If this buffer completes a valid TLS Client Hello, it returns errTLSClientHelloFullyReceived.
+// If an invalid TLS Client Hello message is detected, it returns the error errInvalidTLSClientHello.
+// If this buffer still requires more data to build a complete TLS Client Hello message, it returns nil error.
+//
+// You can call ReadFrom multiple times if r doesn't provide enough data to build a complete Client Hello packet.
+func (b *clientHelloBuffer) ReadFrom(r io.Reader) (n int64, err error) {
+	if !b.valid {
+		return 0, errInvalidTLSClientHello
+	}
+
+	for len(b.data) < cap(b.data)-b.padding && err == nil {
+		m, e := r.Read(b.data[len(b.data) : cap(b.data)-b.padding])
+		b.data = b.data[:len(b.data)+m]
+		n += int64(m)
+		err = e
```

Comment:
Because according to the doc https://pkg.go.dev/io#Reader:

> Callers should always process the n > 0 bytes returned before considering the error err.

In our case, "processing" means determine whether we have read a full header or a full record. After that, we will jump out of the loop when `err != nil` (see the condition in `for` loop).